
<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Lubbock</font></h1>

## Introduction

In this project, the neighborhoods in the city of Lubbock is explored, segmented, and clustered. Data is scraped the city of Lubbock website, the Lubbock area connect website and the Geonames website

Data from the website pages was scraped, wrangled, cleaned and then read it into an excel file and pandas dataframe so that it is in a structured format.

Once the data is in a structured format, the analysis is to explore and cluster the neighborhoods in the city of Lubbock.

## Project Description  
Lubbock is a city in Texas with approximately 307,500 residents living in the area. As at 2017, according to the US Bureau of  Statistics, the total employer establishments are 7,202 which shows an increasing trend. Majority of these businesses include service companies and small businesses. A car salesman is looking to invest in his business and expand to Lubbock. With the increasing number of student population at Texas Tech University, there seems to be more opportunities for a car sales business. Using this project, an analysis scan be done to show the features of the Lubbock neighborhood and where the proposed car business can be established, 
The project requires geo-locational information about Lubbock, Texas can provide the salesman data needed to make his decision. The project will help stakeholders including other car salespersons, parents/students/families in Lubbock, other small businesses and the city of Lubbock as a whole. 
To illustrate, the project will compare the different zip codes in Lubbock and analyze the neighborhoods with the most common car shops and where there are no car shops. Using the k-means clustering algorithm, this will help give a better understanding of the neighborhood and insights on areas to establish the car shop.  

##Data sets and APIs
The data for Lubbock neighborhood that include the zip codes was found on the city of Lubbock website, the Lubbock area connect website and the Geonames website. The major data scraped from these websites include the zip codes, the location data (longitude and latitude) and the county names. 
After data was downloaded from these three websites, the data was joined into one table. There was no missing data based on the zip codes obtained.  After cleaning the data, there were a total of 33 samples and 4 features. The features include: zip code, county name, longitude and latitude. 
The Foursquare API is a data gathering source that is also used for this project as it has a very large database which gives ability to share business locations and provide information based on location search that will be used to understand businesses in Lubbock. Photos and reviews by users provided by Foursquare API can also be used for gathering insights on car shops in Lubbock.
Python scientific libraries, visualization libraries, packages and dependences will also be utilized in getting information about the city of Lubbock. The k-mean clustering algorithm will be applied on the clusters of categories in the Lubbock neighborhood. 




## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>

2. <a href="#item2">Explore Neighborhoods in Lubbock</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c anaconda xlrd --yes #!conda install -c anaconda xlrd

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
!pip install beautifulsoup4
from bs4 import BeautifulSoup

!pip install wikipedia
import wikipedia as wp

!pip install lxml
import lxml as lh

print('Libraries imported.')


Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - xlrd


The following packages will be UPDATED:

    certifi: 2019.11.28-py36_0 conda-forge --> 2019.11.28-py36_0 anaconda
    openssl: 1.1.1d-h516909a_0 conda-forge --> 1.1.1-h7b6447c_0  anaconda

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be UPDATED:

    certifi: 2019.11.28-py

<a id='item1'></a>

## 1. Download and Explore Dataset

Data is scraped, downloaded and extracted

In [2]:
#download data from geonames
import requests
import pandas as pd

url = 'https://www.geonames.org/postal-codes/US/TX/303/lubbock.html'
html = requests.get(url).content
df_list = pd.read_html(html)
df1 = df_list[-1]
print(df1)
df1.to_csv('my data.csv')

    Unnamed: 0            Place             Code          Country  \
0          1.0          Lubbock            79401    United States   
1          NaN  33.587/-101.861  33.587/-101.861  33.587/-101.861   
2          2.0          Lubbock            79404    United States   
3          NaN  33.526/-101.833  33.526/-101.833  33.526/-101.833   
4          3.0          Lubbock            79407    United States   
5          NaN  33.568/-101.942  33.568/-101.942  33.568/-101.942   
6          4.0          Lubbock            79408    United States   
7          NaN  33.566/-101.927  33.566/-101.927  33.566/-101.927   
8          5.0          Lubbock            79410    United States   
9          NaN   33.569/-101.89   33.569/-101.89   33.569/-101.89   
10         6.0          Lubbock            79412    United States   
11         NaN  33.546/-101.858  33.546/-101.858  33.546/-101.858   
12         7.0          Lubbock            79414    United States   
13         NaN   33.55/-101.919   

In [3]:
#download data from area connect
import requests
import pandas as pd

url = 'http://lubbock.areaconnect.com/zip2.htm?city=Lubbock&qs=TX&searchtype=bycity'
html = requests.get(url).content
df_list2 = pd.read_html(html)
df2 = df_list2[-4]
print(df2)
df2.to_csv('my data2.csv')


                                 0
0  Lubbock Zip Code Search Results


In [4]:
df1


Unnamed: 0            Place             Code          Country  \
0          1.0          Lubbock            79401    United States   
1          NaN  33.587/-101.861  33.587/-101.861  33.587/-101.861   
2          2.0          Lubbock            79404    United States   
3          NaN  33.526/-101.833  33.526/-101.833  33.526/-101.833   
4          3.0          Lubbock            79407    United States   
5          NaN  33.568/-101.942  33.568/-101.942  33.568/-101.942   
6          4.0          Lubbock            79408    United States   
7          NaN  33.566/-101.927  33.566/-101.927  33.566/-101.927   
8          5.0          Lubbock            79410    United States   
9          NaN   33.569/-101.89   33.569/-101.89   33.569/-101.89   
10         6.0          Lubbock            79412    United States   
11         NaN  33.546/-101.858  33.546/-101.858  33.546/-101.858   
12         7.0          Lubbock            79414    United States   
13         NaN   33.55/-101.919   33.55/-101.919   33.55/-101.919   
14         8.0          Lubbock            79415    United States   
15         NaN  33.602/-101.876  33.602/-101.876  33.602/-101.876   
16         9.0          Lubbock            79423    United States   
17         NaN  33.515/-101.879  33.515/-101.879  33.515/-101.879   
18        10.0          Lubbock            79424    United States   
19         NaN  33.516/-101.934  33.516/-101.934  33.516/-101.934   
20        11.0           Idalou            79329    United States   
21         NaN   33.65/-101.679   33.65/-101.679   33.65/-101.679   
22        12.0         New Deal            79350    United States   
23         NaN   33.73/-101.836   33.73/-101.836   33.73/-101.836   
24        13.0      Shallowater            79363    United States   
25         NaN  33.692/-101.984  33.692/-101.984  33.692/-101.984   
26        14.0           Slaton            79364    United States   
27         NaN  33.437/-101.643  33.437/-101.643  33.437/-101.643   
28        15.0    Ransom Canyon            79366    United States   
29         NaN   33.533/-101.68   33.533/-101.68   33.533/-101.68   
30        16.0        Wolfforth            79382    United States   
31         NaN  33.506/-102.009  33.506/-102.009  33.506/-102.009   
32        17.0          Lubbock            79402    United States   
33         NaN  33.592/-101.851  33.592/-101.851  33.592/-101.851   
34        18.0          Lubbock            79403    United States   
35         NaN    33.62/-101.81    33.62/-101.81    33.62/-101.81   
36        19.0          Lubbock            79406    United States   
37         NaN  33.582/-101.878  33.582/-101.878  33.582/-101.878   
38        20.0          Lubbock            79409    United States   
39         NaN  33.578/-101.855  33.578/-101.855  33.578/-101.855   
40        21.0          Lubbock            79411    United States   
41         NaN   33.57/-101.863   33.57/-101.863   33.57/-101.863   
42        22.0          Lubbock            79413    United States   
43         NaN  33.547/-101.887  33.547/-101.887  33.547/-101.887   
44        23.0          Lubbock            79416    United States   
45         NaN  33.592/-101.937  33.592/-101.937  33.592/-101.937   
46        24.0          Lubbock            79430    United States   
47         NaN   33.61/-101.821   33.61/-101.821   33.61/-101.821   
48        25.0          Lubbock            79452    United States   
49         NaN   33.61/-101.821   33.61/-101.821   33.61/-101.821   
50        26.0          Lubbock            79453    United States   
51         NaN  33.578/-101.855  33.578/-101.855  33.578/-101.855   
52        27.0          Lubbock            79464    United States   
53         NaN   33.49/-102.011   33.49/-102.011   33.49/-102.011   
54        28.0          Lubbock            79490    United States   
55         NaN   33.61/-101.821   33.61/-101.821   33.61/-101.821   
56        29.0          Lubbock            79493    United States   
57

In [5]:
#Merge dataframes to contain zipcode, county and location data

df = pd.read_excel(r'lubbockcodes.xlsx')

print('Data downloaded and read into a dataframe!')

Data downloaded and read into a dataframe!


In [6]:
df

Zip Code      City Name  Latitude  Longitude
0      79401   Lubbock79401    33.585   -101.847
1      79402   Lubbock79402    33.588   -101.847
2      79403   Lubbock79403    33.603   -101.815
3      79404   Lubbock79404    33.554   -101.821
4      79405   Lubbock79405    33.572   -101.850
5      79406   Lubbock79406    33.580   -101.880
6      79407   Lubbock79407    33.570   -101.948
7      79408   Lubbock79408    33.581   -101.978
8      79409   Lubbock79409    33.586   -101.854
9      79410   Lubbock79410    33.568   -101.895
10     79411   Lubbock79411    33.570   -101.861
11     79412   Lubbock79412    33.545   -101.855
12     79413   Lubbock79413    33.544   -101.890
13     79414   Lubbock79414    33.549   -101.923
14     79415   Lubbock79415    33.606   -101.874
15     79416   Lubbock79416    33.590   -101.934
16     79423   Lubbock79423    33.515   -101.883
17     79424   Lubbock79424    33.523   -101.927
18     79430   Lubbock79430    33.593   -101.899
19     79452   Lubbock79452    33.549   -101.848
20     79453   Lubbock79453    33.577   -101.854
21     79457   Lubbock79457    33.585   -101.852
22     79464   Lubbock79464    33.536   -101.916
23     79490   Lubbock79490    33.577   -101.983
24     79491   Lubbock79491    33.504   -101.878
25     79493   Lubbock79493    33.548   -101.885
26     79499   Lubbock79499    33.608   -101.907
27     79329         Idalou    33.650   -101.679
28     79350       New Deal    33.730   -101.836
29     79363    Shallowater    33.692   -101.984
30     79364         Slaton    33.437   -101.643
31     79366  Ransom Canyon    33.533   -101.680
32     79382     Wolfforth     33.506   -102.009

In [7]:
# let's rename the columns so that they make sense
df.rename(columns={'City Name':'County'}, inplace=True)

In [8]:
df.head()

Zip Code        County  Latitude  Longitude
0     79401  Lubbock79401    33.585   -101.847
1     79402  Lubbock79402    33.588   -101.847
2     79403  Lubbock79403    33.603   -101.815
3     79404  Lubbock79404    33.554   -101.821
4     79405  Lubbock79405    33.572   -101.850

In [9]:
df.shape

(33, 4)

#### Use geopy library to get the latitude and longitude values of Lubbock.

In [10]:
address = 'Lubbock'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Lubbock are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Lubbock are 33.5778631, -101.8551665.


#### Create a map of Lubbock with neighborhoods superimposed on top.

In [11]:
# create map of Lubbock using latitude and longitude values
map_lubbock = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Zip Code'], df['County']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_lubbock)  
    
map_lubbock

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

Explore and cluster the neighborhoods in Lubbock.

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [13]:
df.loc[0, 'Zip Code']

79401

Get the neighborhood's latitude and longitude values.

In [14]:
neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Zip Code'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of 79401 are 33.585, -101.847.


#### Now, let's get the top 100 venues that are in 79401 within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [ ]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL


Send the GET request and examine the resutls

In [16]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5de80a68edbcad001b83ce9c'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Downtown Lubbock',
  'headerFullLocation': 'Downtown Lubbock, Lubbock',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 6,
  'suggestedBounds': {'ne': {'lat': 33.5895000045, 'lng': -101.84160834796657},
   'sw': {'lat': 33.5804999955, 'lng': -101.85239165203342}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b16ac1af964a520ccbb23e3',
       'name': 'Giorgios Pizza',
       'location': {'address': '1020 Ave J',
        'lat': 33.584780488465334,
        'lng': -101.84666292196528,
        'labeledLatLngs': [{'label': 'display',
          'lat': 33.584

In [17]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [18]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name       categories        lat         lng
0         Giorgios Pizza      Pizza Place  33.584780 -101.846663
1         The West Table           Bistro  33.584847 -101.848237
2            Wells Fargo             Bank  33.584911 -101.851885
3               Levine's      Men's Store  33.584292 -101.846655
4  Mcwhorter Tire & Auto  Automotive Shop  33.586174 -101.845960

In [134]:
#Determine the number of venues returned by Foursquare

In [19]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

6 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Lubbock

In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Zip Code', 
                  'Zip Code Latitude', 
                  'Zip Code Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
lubbock_venues = getNearbyVenues(names=df['Zip Code'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

79401
79402
79403
79404
79405
79406
79407
79408
79409
79410
79411
79412
79413
79414
79415
79416
79423
79424
79430
79452
79453
79457
79464
79490
79491
79493
79499
79329
79350
79363
79364
79366
79382


#### Let's check the size of the resulting dataframe

In [22]:
print(lubbock_venues.shape)
lubbock_venues.head()

(242, 7)


Zip Code  Zip Code Latitude  Zip Code Longitude                  Venue  \
0     79401             33.585            -101.847         Giorgios Pizza   
1     79401             33.585            -101.847         The West Table   
2     79401             33.585            -101.847            Wells Fargo   
3     79401             33.585            -101.847               Levine's   
4     79401             33.585            -101.847  Mcwhorter Tire & Auto   

   Venue Latitude  Venue Longitude   Venue Category  
0       33.584780      -101.846663      Pizza Place  
1       33.584847      -101.848237           Bistro  
2       33.584911      -101.851885             Bank  
3       33.584292      -101.846655      Men's Store  
4       33.586174      -101.845960  Automotive Shop

Let's check how many venues were returned for each neighborhood

In [23]:
lubbock_venues.groupby('Zip Code').count()

Zip Code Latitude  Zip Code Longitude  Venue  Venue Latitude  \
Zip Code                                                                 
79364                     4                   4      4               4   
79366                     1                   1      1               1   
79382                     3                   3      3               3   
79401                     6                   6      6               6   
79402                     7                   7      7               7   
79404                     4                   4      4               4   
79405                     6                   6      6               6   
79406                    12                  12     12              12   
79407                    27                  27     27              27   
79408                     2                   2      2               2   
79409                    13                  13     13              13   
79410                     2                   2      2               2   
79411                     5                   5      5               5   
79412                    17                  17     17              17   
79413                     6                   6      6               6   
79414                    30                  30     30              30   
79415                     4                   4      4               4   
79416                    20                  20     20              20   
79424                     4                   4      4               4   
79430                     1                   1      1               1   
79452                    14                  14     14              14   
79453                    10                  10     10              10   
79457                     8                   8      8               8   
79464                    14                  14     14              14   
79490                     1                   1      1               1   
79491                     3                   3      3               3   
79493                    14                  14     14              14   
79499                     4                   4      4               4   

          Venue Longitude  Venue Category  
Zip Code                                   
79364                   4               4  
79366                   1               1  
79382                   3               3  
79401                   6               6  
79402                   7               7  
79404                   4               4  
79405                   6               6  
79406                  12              12  
79407                  27              27  
79408                   2               2  
79409                  13              13  
79410                   2               2  
79411                   5               5  
79412                  17              17  
79413                   6               6  
79414                  30              30  
79415                   4               4  
79416                  20              20  
79424                   4               4  
79430                   1               1  
79452                  14              14  
79453                  10              10  
79457                   8               8  
79464                  14              14  
79490                   1               1  
79491                   3               3  
79493                  14              14  
79499                   4               4

#### Let's find out how many unique categories can be curated from all the returned venues

In [24]:
print('There are {} uniques categories.'.format(len(lubbock_venues['Venue Category'].unique())))

There are 102 uniques categories.


In [25]:
# one hot encoding
lubbock_onehot = pd.get_dummies(lubbock_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
lubbock_onehot['Zip Code'] = lubbock_venues['Zip Code'] 

# move neighborhood column to the first column
fixed_columns = [lubbock_onehot.columns[-1]] + list(lubbock_onehot.columns[:-1])
lubbock_onehot = lubbock_onehot[fixed_columns]

lubbock_onehot.head()


Zip Code  ATM  Accessories Store  American Restaurant  Antique Shop  \
0     79401    0                  0                    0             0   
1     79401    0                  0                    0             0   
2     79401    0                  0                    0             0   
3     79401    0                  0                    0             0   
4     79401    0                  0                    0             0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 0                   0   
4            0                    0                 0                   0   

   Automotive Shop  BBQ Joint  Bakery  Bank  Bar  Baseball Field  \
0                0          0       0     0    0               0   
1                0          0       0     0    0               0   
2                0          0       0     1    0               0   
3                0          0       0     0    0               0   
4                1          0       0     0    0               0   

   Basketball Court  Big Box Store  Bistro  Bookstore  Brewery  Bridal Shop  \
0                 0              0       0          0        0            0   
1                 0              0       1          0        0            0   
2                 0              0       0          0        0            0   
3                 0              0       0          0        0            0   
4                 0              0       0          0        0            0   

   Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Business Service  \
0                0       0             0              0                 0   
1                0       0             0              0                 0   
2                0       0             0              0                 0   
3                0       0             0              0                 0   
4                0       0             0              0                 0   

   Cajun / Creole Restaurant  Chinese Restaurant  Clothing Store  Coffee Shop  \
0                          0                   0               0            0   
1                          0                   0               0            0   
2                          0                   0               0            0   
3                          0                   0               0            0   
4                          0                   0               0            0   

   College Bookstore  College Cafeteria  College Quad  College Residence Hall  \
0                  0                  0             0                       0   
1                  0                  0             0                       0   
2                  0                  0             0                       0   
3                  0                  0             0                       0   
4                  0                  0             0                       0   

   College Theater  Convenience Store  Cosmetics Shop  Department Store  \
0                0                  0               0                 0   
1                0                  0               0                 0   
2                0                  0               0                 0   
3                0                  0               0                 0   
4                0                  0               0                 0   

   Dessert Shop  Diner  Discount Store  Dive Bar  Electronics Store  Farm  \
0             0      0               0         0                  0     0   
1             0      0               0         0                  0     0   
2             0      0               0         0                  0     0   
3             0

And let's examine the new dataframe size.

In [26]:
lubbock_onehot.shape

(242, 103)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [27]:
lubbock_grouped = lubbock_onehot.groupby('Zip Code').mean().reset_index()
lubbock_grouped

Zip Code       ATM  Accessories Store  American Restaurant  Antique Shop  \
0      79364  0.000000           0.000000             0.000000           0.0   
1      79366  0.000000           0.000000             0.000000           0.0   
2      79382  0.000000           0.000000             0.000000           0.0   
3      79401  0.000000           0.000000             0.000000           0.0   
4      79402  0.000000           0.000000             0.000000           0.0   
5      79404  0.000000           0.000000             0.000000           0.0   
6      79405  0.000000           0.000000             0.000000           0.0   
7      79406  0.000000           0.000000             0.000000           0.0   
8      79407  0.000000           0.037037             0.000000           0.0   
9      79408  0.000000           0.000000             0.000000           0.0   
10     79409  0.000000           0.000000             0.000000           0.0   
11     79410  0.000000           0.000000             0.000000           0.0   
12     79411  0.000000           0.000000             0.000000           0.0   
13     79412  0.000000           0.000000             0.117647           0.0   
14     79413  0.000000           0.000000             0.000000           0.0   
15     79414  0.033333           0.000000             0.033333           0.0   
16     79415  0.000000           0.000000             0.000000           0.0   
17     79416  0.000000           0.000000             0.050000           0.0   
18     79424  0.000000           0.000000             0.000000           0.0   
19     79430  0.000000           0.000000             0.000000           0.0   
20     79452  0.071429           0.000000             0.071429           0.0   
21     79453  0.000000           0.000000             0.000000           0.0   
22     79457  0.000000           0.000000             0.000000           0.0   
23     79464  0.000000           0.000000             0.071429           0.0   
24     79490  0.000000           0.000000             0.000000           1.0   
25     79491  0.000000           0.000000             0.000000           0.0   
26     79493  0.000000           0.000000             0.000000           0.0   
27     79499  0.000000           0.000000             0.000000           0.0   

    Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0      0.000000                  0.0          0.000000            0.000000   
1      0.000000                  0.0          0.000000            0.000000   
2      0.000000                  0.0          0.000000            0.000000   
3      0.000000                  0.0          0.000000            0.000000   
4      0.428571                  0.0          0.000000            0.000000   
5      0.000000                  0.0          0.000000            0.000000   
6      0.000000                  0.0          0.000000            0.000000   
7      0.083333                  0.0          0.000000            0.000000   
8      0.000000                  0.0          0.000000            0.000000   
9      0.000000                  0.5          0.000000            0.000000   
10     0.000000                  0.0          0.000000            0.000000   
11     0.000000                  0.0          0.000000            0.000000   
12     0.000000                  0.0          0.000000            0.000000   
13     0.000000                  0.0          0.000000            0.000000   
14     0.000000                  0.0          0.000000            0.000000   
15     0.000000                  0.0          0.033333            0.000000   
16     0.000000                  0.0          0.000000            0.000000   
17     0.000000                  0.0          0.000000            0.000000   
18     0.000000                  0.0          0.000000            0.000000   
19     0.000000                  0.0          0.000000            0.000000   
20     0.000000                  0.0         

#### Let's confirm the new size


In [28]:
lubbock_grouped.shape


(28, 103)

In [29]:
lubbock_grouped.head()

Zip Code  ATM  Accessories Store  American Restaurant  Antique Shop  \
0     79364  0.0                0.0                  0.0           0.0   
1     79366  0.0                0.0                  0.0           0.0   
2     79382  0.0                0.0                  0.0           0.0   
3     79401  0.0                0.0                  0.0           0.0   
4     79402  0.0                0.0                  0.0           0.0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0     0.000000                  0.0               0.0                 0.0   
1     0.000000                  0.0               0.0                 0.0   
2     0.000000                  0.0               0.0                 0.0   
3     0.000000                  0.0               0.0                 0.0   
4     0.428571                  0.0               0.0                 0.0   

   Automotive Shop  BBQ Joint    Bakery      Bank  Bar  Baseball Field  \
0         0.000000       0.25  0.250000  0.000000  0.0             0.0   
1         0.000000       0.00  0.000000  0.000000  0.0             0.0   
2         0.000000       0.00  0.000000  0.000000  0.0             0.0   
3         0.166667       0.00  0.166667  0.166667  0.0             0.0   
4         0.142857       0.00  0.000000  0.000000  0.0             0.0   

   Basketball Court  Big Box Store    Bistro  Bookstore  Brewery  Bridal Shop  \
0               0.0            0.0  0.000000        0.0      0.0          0.0   
1               0.0            0.0  0.000000        0.0      0.0          0.0   
2               0.0            0.0  0.000000        0.0      0.0          0.0   
3               0.0            0.0  0.166667        0.0      0.0          0.0   
4               0.0            0.0  0.142857        0.0      0.0          0.0   

   Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Business Service  \
0              0.0     0.0           0.0            0.0               0.0   
1              0.0     0.0           0.0            0.0               0.0   
2              0.0     0.0           0.0            0.0               0.0   
3              0.0     0.0           0.0            0.0               0.0   
4              0.0     0.0           0.0            0.0               0.0   

   Cajun / Creole Restaurant  Chinese Restaurant  Clothing Store  Coffee Shop  \
0                        0.0                 0.0             0.0          0.0   
1                        0.0                 0.0             0.0          0.0   
2                        0.0                 0.0             0.0          0.0   
3                        0.0                 0.0             0.0          0.0   
4                        0.0                 0.0             0.0          0.0   

   College Bookstore  College Cafeteria  College Quad  College Residence Hall  \
0                0.0                0.0           0.0                     0.0   
1                0.0                0.0           0.0                     0.0   
2                0.0                0.0           0.0                     0.0   
3                0.0                0.0           0.0                     0.0   
4                0.0                0.0           0.0                     0.0   

   College Theater  Convenience Store  Cosmetics Shop  Department Store  \
0              0.0               0.25             0.0               0.0   
1              0.0               0.00             0.0               0.0   
2              0.0               0.00             0.0               0.0   
3              0.0               0.00             0.0               0.0   
4              0.0               0.00             0.0               0.0   

   Dessert Shop  Diner  Discount Store  Dive Bar  Electronics Store  Farm  \
0           0.0    0.0        0.000000       0.0                0.0   0.0   
1           0.0    0.0        0.000000       0.0                0.0   0.0   
2           0.0    0.0        0.333333      

In [30]:
lubbock_grouped.dtypes

Zip Code                       int64
ATM                          float64
Accessories Store            float64
American Restaurant          float64
Antique Shop                 float64
Art Gallery                  float64
Arts & Crafts Store          float64
Asian Restaurant             float64
Athletics & Sports           float64
Automotive Shop              float64
BBQ Joint                    float64
Bakery                       float64
Bank                         float64
Bar                          float64
Baseball Field               float64
Basketball Court             float64
Big Box Store                float64
Bistro                       float64
Bookstore                    float64
Brewery                      float64
Bridal Shop                  float64
Bubble Tea Shop              float64
Buffet                       float64
Burger Joint                 float64
Burrito Place                float64
Business Service             float64
Cajun / Creole Restaurant    float64
C

#### Let's print each neighborhood along with the top 5 most common venues

In [31]:
num_top_venues = 5

for hood in lubbock_grouped['Zip Code']:
    print(hood)
    temp = lubbock_grouped[lubbock_grouped['Zip Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

79364
               venue  freq
0  Convenience Store  0.25
1        Flower Shop  0.25
2          BBQ Joint  0.25
3             Bakery  0.25
4                ATM  0.00


79366
                  venue  freq
0        Scenic Lookout   1.0
1                   ATM   0.0
2           Pizza Place   0.0
3                  Park   0.0
4  Outdoor Supply Store   0.0


79382
                  venue  freq
0        Discount Store  0.33
1         Train Station  0.33
2        Sandwich Place  0.33
3    Italian Restaurant  0.00
4  Outdoor Supply Store  0.00


79401
             venue  freq
0      Pizza Place  0.17
1           Bistro  0.17
2  Automotive Shop  0.17
3           Bakery  0.17
4             Bank  0.17


79402
             venue  freq
0      Art Gallery  0.43
1      Pizza Place  0.14
2  Automotive Shop  0.14
3      Men's Store  0.14
4           Bistro  0.14


79404
            venue  freq
0  Hardware Store  0.25
1    Food Service  0.25
2            Food  0.25
3  Sandwich Place  0.25
4           

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [32]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [33]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Zip Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Zip Code'] = lubbock_grouped['Zip Code']

for ind in np.arange(lubbock_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(lubbock_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Zip Code 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0     79364           Flower Shop     Convenience Store             BBQ Joint   
1     79366        Scenic Lookout         Women's Store      Business Service   
2     79382         Train Station        Sandwich Place        Discount Store   
3     79401           Pizza Place           Men's Store       Automotive Shop   
4     79402           Art Gallery           Pizza Place           Men's Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                Bakery         Women's Store          Dessert Shop   
1    Chinese Restaurant        Clothing Store           Coffee Shop   
2         Women's Store        Cosmetics Shop    Chinese Restaurant   
3                Bakery                  Bank                Bistro   
4       Automotive Shop                Bistro         Women's Store   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Coffee Shop     College Bookstore     College Cafeteria   
1     College Bookstore     College Cafeteria          College Quad   
2        Clothing Store           Coffee Shop     College Bookstore   
3         Women's Store      Department Store     College Bookstore   
4      Department Store           Coffee Shop     College Bookstore   

   10th Most Common Venue  
0            College Quad  
1  College Residence Hall  
2       College Cafeteria  
3       College Cafeteria  
4       College Cafeteria

In [34]:
neighborhoods_venues_sorted

Zip Code 1st Most Common Venue   2nd Most Common Venue  \
0      79364           Flower Shop       Convenience Store   
1      79366        Scenic Lookout           Women's Store   
2      79382         Train Station          Sandwich Place   
3      79401           Pizza Place             Men's Store   
4      79402           Art Gallery             Pizza Place   
5      79404          Food Service                    Food   
6      79405           Pizza Place             Flea Market   
7      79406                  Park            Food Service   
8      79407        Clothing Store  Furniture / Home Store   
9      79408           Flower Shop     Arts & Crafts Store   
10     79409                  Bank                   Diner   
11     79410     Recreation Center                     Gym   
12     79411      Business Service               Locksmith   
13     79412    Mexican Restaurant                Dive Bar   
14     79413           Coffee Shop          Discount Store   
15     79414  Fast Food Restaurant          Discount Store   
16     79415        Baseball Field         Motorcycle Shop   
17     79416           Pizza Place              Sports Bar   
18     79424  Fast Food Restaurant        Sushi Restaurant   
19     79430  Fast Food Restaurant           Women's Store   
20     79452  Fast Food Restaurant                   Hotel   
21     79453    Mexican Restaurant              Smoke Shop   
22     79457   Fried Chicken Joint             Pizza Place   
23     79464            Steakhouse      Mexican Restaurant   
24     79490          Antique Shop           Women's Store   
25     79491  Gym / Fitness Center              Hobby Shop   
26     79493           Coffee Shop      Salon / Barbershop   
27     79499              Wine Bar             Video Store   

        3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0                   BBQ Joint                  Bakery         Women's Store   
1            Business Service      Chinese Restaurant        Clothing Store   
2              Discount Store           Women's Store        Cosmetics Shop   
3             Automotive Shop                  Bakery                  Bank   
4                 Men's Store         Automotive Shop                Bistro   
5              Hardware Store          Sandwich Place      Department Store   
6        Fast Food Restaurant  Thrift / Vintage Store       Thai Restaurant   
7                        Pool             Art Gallery                  Farm   
8   Cajun / Creole Restaurant           Women's Store     Mobile Phone Shop   
9               Women's Store            Dessert Shop        Clothing Store   
10                   Pharmacy      Italian Restaurant     Convenience Store   
11              Women's Store        Department Store        Clothing Store   
12          Convenience Store             IT Services                  Park   
13        American Restaurant                Pharmacy           Pizza Place   
14                Supermarket         Automotive Shop                Bakery   
15                        Bar                     ATM    Seafood Restaurant   
16               Burger Joint                 Brewery         Women's Store   
17              Big Box Store     Rental Car Location        Cosmetics Shop   
18                  BBQ Joint          Cosmetics Shop         Women's Store   
19               Dessert Shop          Clothing Store           Coffee Shop   
20         Mexican Restaurant                     ATM        Sandwich Place   
21          Convenience Store  Furniture / Home Store         Burrito Place   
22              Hot Dog Joint                   Hotel     Convenience Store   
23                  BBQ Joint              Shoe Store           Men's Store   
24               Dessert Shop          Clothing Store           Coffee Shop   
25         Athletics & Sports           Women's Store          Dessert Shop   
26                Supermarket            Intersection          Dessert S

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [35]:
# set number of clusters
kclusters = 5

lubbock_grouped_clustering = lubbock_grouped.drop('Zip Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(lubbock_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 2, 1, 0, 1, 1, 0, 1, 0, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [36]:
# add clustering labels
neighborhoods_venues_sorted.insert(0,'Label Cluster', kmeans.labels_)

lubbock_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
lubbock_merged = lubbock_merged.join(neighborhoods_venues_sorted.set_index('Zip Code'), on='Zip Code')

lubbock_merged.head() # check the last columns!

Zip Code        County  Latitude  Longitude  Label Cluster  \
0     79401  Lubbock79401    33.585   -101.847            0.0   
1     79402  Lubbock79402    33.588   -101.847            1.0   
2     79403  Lubbock79403    33.603   -101.815            NaN   
3     79404  Lubbock79404    33.554   -101.821            1.0   
4     79405  Lubbock79405    33.572   -101.850            0.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Pizza Place           Men's Store       Automotive Shop   
1           Art Gallery           Pizza Place           Men's Store   
2                   NaN                   NaN                   NaN   
3          Food Service                  Food        Hardware Store   
4           Pizza Place           Flea Market  Fast Food Restaurant   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                  Bakery                  Bank                Bistro   
1         Automotive Shop                Bistro         Women's Store   
2                     NaN                   NaN                   NaN   
3          Sandwich Place      Department Store        Clothing Store   
4  Thrift / Vintage Store       Thai Restaurant                   Bar   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0         Women's Store      Department Store     College Bookstore   
1      Department Store           Coffee Shop     College Bookstore   
2                   NaN                   NaN                   NaN   
3           Coffee Shop     College Bookstore     College Cafeteria   
4         Women's Store     Convenience Store    Chinese Restaurant   

  10th Most Common Venue  
0      College Cafeteria  
1      College Cafeteria  
2                    NaN  
3           College Quad  
4         Clothing Store

In [37]:
#convert clusters
lubbock_merged["Label Cluster"] = lubbock_merged["Label Cluster"].fillna("0").astype(int)

In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(lubbock_merged['Latitude'], lubbock_merged['Longitude'], lubbock_merged['Zip Code'], lubbock_merged['Label Cluster']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, each cluster is assigned a name

#### Cluster 1

In [39]:
lubbock_merged.loc[lubbock_merged['Label Cluster'] == 0, lubbock_merged.columns[[1] + list(range(5, lubbock_merged.shape[1]))]]

County 1st Most Common Venue   2nd Most Common Venue  \
0   Lubbock79401           Pizza Place             Men's Store   
2   Lubbock79403                   NaN                     NaN   
4   Lubbock79405           Pizza Place             Flea Market   
6   Lubbock79407        Clothing Store  Furniture / Home Store   
8   Lubbock79409                  Bank                   Diner   
11  Lubbock79412    Mexican Restaurant                Dive Bar   
12  Lubbock79413           Coffee Shop          Discount Store   
13  Lubbock79414  Fast Food Restaurant          Discount Store   
15  Lubbock79416           Pizza Place              Sports Bar   
16  Lubbock79423                   NaN                     NaN   
17  Lubbock79424  Fast Food Restaurant        Sushi Restaurant   
19  Lubbock79452  Fast Food Restaurant                   Hotel   
20  Lubbock79453    Mexican Restaurant              Smoke Shop   
21  Lubbock79457   Fried Chicken Joint             Pizza Place   
22  Lubbock79464            Steakhouse      Mexican Restaurant   
25  Lubbock79493           Coffee Shop      Salon / Barbershop   
26  Lubbock79499              Wine Bar             Video Store   
27        Idalou                   NaN                     NaN   
28      New Deal                   NaN                     NaN   
29   Shallowater                   NaN                     NaN   

        3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0             Automotive Shop                  Bakery                  Bank   
2                         NaN                     NaN                   NaN   
4        Fast Food Restaurant  Thrift / Vintage Store       Thai Restaurant   
6   Cajun / Creole Restaurant           Women's Store     Mobile Phone Shop   
8                    Pharmacy      Italian Restaurant     Convenience Store   
11        American Restaurant                Pharmacy           Pizza Place   
12                Supermarket         Automotive Shop                Bakery   
13                        Bar                     ATM    Seafood Restaurant   
15              Big Box Store     Rental Car Location        Cosmetics Shop   
16                        NaN                     NaN                   NaN   
17                  BBQ Joint          Cosmetics Shop         Women's Store   
19         Mexican Restaurant                     ATM        Sandwich Place   
20          Convenience Store  Furniture / Home Store         Burrito Place   
21              Hot Dog Joint                   Hotel     Convenience Store   
22                  BBQ Joint              Shoe Store           Men's Store   
25                Supermarket            Intersection          Dessert Shop   
26          Convenience Store            Intersection         Women's Store   
27                        NaN                     NaN                   NaN   
28                        NaN                     NaN                   NaN   
29                        NaN                     NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Bistro         Women's Store      Department Store   
2                    NaN                   NaN                   NaN   
4                    Bar         Women's Store     Convenience Store   
6     Mexican Restaurant        Hunting Supply     Electronics Store   
8                  Hotel         Hot Dog Joint                   Spa   
11       Bubble Tea Shop        Ice Cream Shop                  Bank   
12          Dessert Shop        Clothing Store     College Bookstore   
13    Chinese Restaurant    Mexican Restaurant        Hunting Supply   
15        Sandwich Place          Burger Joint           Gas Station   
16                   NaN                   NaN                   NaN   
17      Department Store        Clothing Store           Coffee Shop   
19            Steakhouse            Hotel Pool   American Restaurant   
20  Fast Food Restaurant   

#### Cluster 2

In [40]:
lubbock_merged.loc[lubbock_merged['Label Cluster'] == 1, lubbock_merged.columns[[1] + list(range(5, lubbock_merged.shape[1]))]]

County 1st Most Common Venue 2nd Most Common Venue  \
1   Lubbock79402           Art Gallery           Pizza Place   
3   Lubbock79404          Food Service                  Food   
5   Lubbock79406                  Park          Food Service   
7   Lubbock79408           Flower Shop   Arts & Crafts Store   
9   Lubbock79410     Recreation Center                   Gym   
10  Lubbock79411      Business Service             Locksmith   
14  Lubbock79415        Baseball Field       Motorcycle Shop   
24  Lubbock79491  Gym / Fitness Center            Hobby Shop   
30        Slaton           Flower Shop     Convenience Store   
32    Wolfforth          Train Station        Sandwich Place   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1            Men's Store       Automotive Shop                Bistro   
3         Hardware Store        Sandwich Place      Department Store   
5                   Pool           Art Gallery                  Farm   
7          Women's Store          Dessert Shop        Clothing Store   
9          Women's Store      Department Store        Clothing Store   
10     Convenience Store           IT Services                  Park   
14          Burger Joint               Brewery         Women's Store   
24    Athletics & Sports         Women's Store          Dessert Shop   
30             BBQ Joint                Bakery         Women's Store   
32        Discount Store         Women's Store        Cosmetics Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1          Women's Store      Department Store           Coffee Shop   
3         Clothing Store           Coffee Shop     College Bookstore   
5            Coffee Shop     College Bookstore     College Cafeteria   
7            Coffee Shop     College Bookstore     College Cafeteria   
9            Coffee Shop     College Bookstore     College Cafeteria   
10        Cosmetics Shop    Chinese Restaurant        Clothing Store   
14           Coffee Shop     College Bookstore     College Cafeteria   
24           Coffee Shop     College Bookstore     College Cafeteria   
30          Dessert Shop           Coffee Shop     College Bookstore   
32    Chinese Restaurant        Clothing Store           Coffee Shop   

   9th Most Common Venue  10th Most Common Venue  
1      College Bookstore       College Cafeteria  
3      College Cafeteria            College Quad  
5           College Quad          Sandwich Place  
7           College Quad  College Residence Hall  
9           College Quad  College Residence Hall  
10           Coffee Shop       College Bookstore  
14          College Quad  College Residence Hall  
24          College Quad  College Residence Hall  
30     College Cafeteria            College Quad  
32     College Bookstore       College Cafeteria

#### Cluster 3

In [41]:
lubbock_merged.loc[lubbock_merged['Label Cluster'] == 2, lubbock_merged.columns[[1] + list(range(5, lubbock_merged.shape[1]))]]

County 1st Most Common Venue 2nd Most Common Venue  \
31  Ransom Canyon        Scenic Lookout         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
31      Business Service    Chinese Restaurant        Clothing Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
31           Coffee Shop     College Bookstore     College Cafeteria   

   9th Most Common Venue  10th Most Common Venue  
31          College Quad  College Residence Hall

#### Cluster 4

In [42]:
lubbock_merged.loc[lubbock_merged['Label Cluster'] == 3, lubbock_merged.columns[[1] + list(range(5, lubbock_merged.shape[1]))]]

County 1st Most Common Venue 2nd Most Common Venue  \
18  Lubbock79430  Fast Food Restaurant         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18          Dessert Shop        Clothing Store           Coffee Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
18     College Bookstore     College Cafeteria          College Quad   

     9th Most Common Venue 10th Most Common Venue  
18  College Residence Hall        College Theater

#### Cluster 5

In [43]:
lubbock_merged.loc[lubbock_merged['Label Cluster'] == 4, lubbock_merged.columns[[1] + list(range(5, lubbock_merged.shape[1]))]]

County 1st Most Common Venue 2nd Most Common Venue  \
23  Lubbock79490          Antique Shop         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
23          Dessert Shop        Clothing Store           Coffee Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
23     College Bookstore     College Cafeteria          College Quad   

     9th Most Common Venue 10th Most Common Venue  
23  College Residence Hall        College Theater

### Thank you. This was done by Olufunke Oladimeji